docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,311 @@
+* **Claim**: A resource that represents a claim on an IP address.
+* **Load Balancer Controller**: a controller capable of fulfilling all Ingress paths for a given class of claims.
+
+__Example__:
+* Cluster is bootstrapped with kube-controller-manager --loadbalancer-controllers=(cloud=gce),haproxy
+* User creates svc1 and svc2
+* User creates claim1 to get a public ip.
+  GCEController:
+    - receives watch notification for `Class: cloud` object.
+    - allocates cloud resources for a new L7
+    - spins off a goroutine that watches services for that one claim and updates the L7
+* User creates claim2.
+  Haproxy Controller:
+    - receives watch notification for `Class: haproxy` object
+    - creates a haproxy pod
+    - haproxy pod watches services for that one claim and reconfigures haproxy
+* We end up with 2 ips:
+    - 172.12.14.111:80 serving foo and bar at L7 through a GCE L7
+    - 172.11.14.143:8080 switching between foobar and baz with different certs using sni at L4, through a pod in the cluster.
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPath
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: www.example.com
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
```

Comment:
Ultimately we said that all public APIs shouldn't use maps
api-conventions.md
<https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#lists-of-named-subobjects-preferred-over-maps>.


On Mon, Aug 24, 2015 at 11:10 PM, Prashanth B <notifications@github.com>
wrote:

> In docs/proposals/loadbalancing.md
> <https://github.com/kubernetes/kubernetes/pull/12827#discussion_r37826454>
> :
>
> > +* Security configuration (i.e tlsMode, secret)
> > +* Metadata needed for more advanced loadbalancing (e.g session persistence)
> > +
> > +See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
> > +
> > +```yaml
> > +apiVersion: v1
> > +kind: IngressPath
> > +metadata:
> > +  name: l7ingress
> > +  type: l7
> > +spec:
> > +  host: www.example.com
> > +  tlsMode: Termination
> > +  secret: foosecret
> > +  pathMap:
>
> Aren't maps verboten? Also, what if I don't want to allow users to pick a
> domain, but give them one? Do they just fill out random values? Array
> fixes both of those problems.
>
> The way I currently have it, it's a map of lists, where the map is for the
> subdomain (
> https://github.com/kubernetes/kubernetes/pull/12827/files#diff-41f2bde570ebc813183b7cd0a96a7e04R137).
> You're not allowed to specify multiple domains per ingresspoint resource
> (which actually makes having the domain at the end there pretty redundant).
> If you want to request a host you'd leave that Spec.Host field blank, and
> one will be assigned and used with your subdomains.
>
> It wouldn't be too hard to do this as a list of lists if you have a use
> case. I used a map because you can't have multiple path regex lists for the
> same subdomain.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12827/files#r37826454>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift